src/runtime/policy/ingress-validation.ts

```diff
@@ -62,41 +62,86 @@ export class IngressValidation {
-        const fieldResults = fieldCapabilities.map(
-            fc => fc.isAllowedForIngress(handle.capabilities));
-        if (!fieldResults.some(r => r.success)) {
-          result.addResult(IngressValidationResult.failWith(handle,
-              `Failed validating ingress for field '${fieldPath}'` +
-              ` of '${handle.id || handle.connections[0].getQualifiedName()}`,
-              fieldResults.filter(r => !r.success)));
+        if (!capabilitiesByField.has(fieldPath)) {
+          capabilitiesByField.set(fieldPath, []);
         }
+        capabilitiesByField.get(fieldPath).push(...fieldCapabilities);
       }
     } else {
-      result.addError(
-        `Handle '${handle.id}' has no matching target type ` +
-            `${handle.type.resolvedType().toString()} in policies`);
+      // All input connections of all particles writing into this handle.
+      const sourceParticles = handle.connections.filter(conn => conn.isOutput)
+          .map(conn => conn.particle);
+      if (sourceParticles.length === 0) {
+        return IngressValidationResult.failWith(handle,
+            `Handle '${handle.id}' has no matching target type ` +
+            `${handle.type.resolvedType().toString()} in policies, and no source particles.`);
+      }
+      const sourceConnections = [];
+      for (const sourceParticle of sourceParticles) {
+        const particleInputs = Object.values(sourceParticle.connections)
+            .filter(conn => conn.isInput && !seenHandles.has(conn.handle));
+        if (particleInputs.length === 0) {
+          return IngressValidationResult.failWith(handle,
+              `Handle '${handle.id}' has no matching target type ` +
+              `${handle.type.resolvedType().toString()} in policies, and ` +
+              `contributing Particle ${sourceParticle.name} has no inputs.`);
+        }
+        sourceConnections.push(...particleInputs);
+      }
+      for (const sourceConn of sourceConnections) {
+        const result = this.findHandleCapabilities(sourceConn.handle, capabilitiesByField);
+        if (!result.success) {
+          return result;
+        }
+      }
+    }
+    return IngressValidationResult.success(handle);
+  }
+
+  // Returns success, if the type of the handle, restricted according with the
+  // set of policies, has capabilities that are compliant with the policies'
+  // retention and maxAge permissions. Otherwise, returns a combination of all
+  // encountered errors.
+  private validateHandleCapabilities(handle: Handle): IngressValidationResult {
+    const capabilitiesByField = new Map<string, Capabilities[]>();
+    const result = this.findHandleCapabilities(handle, capabilitiesByField);
+    if (!result.success) return result;
+    for (const [fieldPath, fieldCapabilities] of capabilitiesByField.entries()) {
+      const fieldResults = fieldCapabilities.map(
+          fc => fc.isAllowedForIngress(handle.capabilities));
+      if (!fieldResults.some(r => r.success)) {
```

Comment:
This says that only one successful result is needed. Is that right? I would have thought that all fields would need to succeed?

Reply:
There is a list of fields.
Each field has a list of Capabilities allowed by policies (because each field may be covered by several policies).
For each field, there must be at least one Capabilities (out of list of Capabilities) that allows the ingress to the handle (and I iterate over all fields in order to collect all errors). 

I added a comment in the code. Hope that makes sense.